rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,473 @@
+device integration until later this year. (For device support expecting to land
+in 2021 or later, we highly recommend waiting to integrate with the new stack,
+since it is fundamentally different from the current stack and cannot guarantee
+code reuse.)
+
+In the meantime, we plan to provide limited device integration support for the
+current TensorFlow stack through
+[Modular TensorFlow](https://github.com/tensorflow/community/blob/master/rfcs/20190305-modular-tensorflow.md).
+We anticipate three major functionalities within a device plugin module:
+
+*   Device registration: Will be addressed in a different RFC.
+*   Device management: The focus of this RFC.
+*   Kernel and op registration and implementation:
+    [RFC Accepted](https://github.com/tensorflow/community/blob/master/rfcs/20190814-kernel-and-op-registration.md). [C API implemented](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/).
+
+[StreamExecutor](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_pimpl.h;l=73) is TensorFlow's main device manager, responsible for work execution and memory management. It provides a set of methods
+(such as
+[Memcpy](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=240))
+that can be customized for a particular device.
+
+We propose a C API wrapper of a subset of methods in
+[StreamExecutorInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=166?q=StreamExecutorinterface)
+as an ABI-stable way to register a custom StreamExecutor platform.
+
+# User Benefits
+
+A decoupled way to add a new device to TensorFlow.
+
+*   Simpler process: Does not have to add a new build toolchain to TensorFlow
+*   Faster time-to-solution: Does not need code review from the TensorFlow team.
+*   Lower maintenance efforts: Only C-API-related changes could break the
+    integration. Unrelated TensorFlow changes would not break the code.
+       *    The C APIs may be changed during the initial experimental phase based 
+            on developer experience and feedback. When the APIs become more mature, 
+            we will try to keep them stable (in a best-effort manner) until the new 
+            TensorFlow stack is available.
+
+# Design Proposal
+
+## StreamExecutorInterface
+
+[StreamExecutorInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=166?q=StreamExecutorinterface)
+is quite large and some of its methods are only sporadically used. Therefore, we
+plan to wrap only a subset of key StreamExecutorInterface functionality.
+
+See proposed C API below:
+
+```cpp
+typedef SE_Stream_st* SE_Stream;
+typedef SE_Event_st* SE_Event;
+typedef SE_Timer_st* SE_Timer;
+typedef TF_Status* (*TF_StatusCallbackFn)(void*);
+
+typedef struct SE_PlatformId {
```

Comment:
Hi, @annarev  I saw there are lots of struct and functions with prefix SE_XXX, like SE_Platform, SE_Stream, some of them are defined as a concrete struct (SE_PlatformId, SE_DeviceDescription) in tensorflow proper, some of them seem like just a handle with typedef (SE_Platform, SE_Stream). is there any description about which are defined in Tensorflow Proper, and which are need to be defined in Plugin? Thanks

Reply:
If we have a pointer-to-name typedef at the top then we treat it as opaque struct created by device plugin. These are opaque types to TensorFlow defined by plugin:
```cpp
typedef SE_Stream_st* SE_Stream;
typedef SE_Event_st* SE_Event;
typedef SE_Timer_st* SE_Timer;
```
`SE_Platform` will be defined by TensorFlow and should be treated as opaque pointer in the plugin.

`SE_PlatformId`, `SE_DeviceDescription`, `SE_TimerFns`, `SE_AllocatorStats`, `SE_EventStatus`, `SE_Options`, `SE_Device`, `SE_StreamExecutor` are all defined as structs in the C API so both sides of ABI boundary can access members.

I agree this is confusing. Let me think if there is a way to differentiate between them (may be with some name suffix).

Reply:
ok. thanks, it is much clear for me now:)

Reply:
I couldn't come up with a good way to add respective annotations to the names, but will keep thinking. At the very list we should provide more documentation when implementing the API.

Reply:
I renamed structs according to the @alextp suggestion from design review:
* `SE_` prefix for structs populated by TensorFlow
* `SP_` prefix for structs populated by plugin

`SE_PlatformId` is removed from C API now. TensorFlow will set it behind the scenes.